src/fdb5/database/Key.h

```diff
@@ -161,7 +163,66 @@ class Key {
     eckit::StringList names_;
 
     std::shared_ptr<TypesRegistry> registry_;
-    bool canonical_;
+};
+
+
+//----------------------------------------------------------------------------------------------------------------------
+
+class CanonicalKey : public Key {
+
+public: // methods
+
+    explicit CanonicalKey(const std::shared_ptr<TypesRegistry> reg = nullptr);
+    explicit CanonicalKey(eckit::Stream &, const std::shared_ptr<TypesRegistry> reg = nullptr);
```

Comment:
I don't understand (a) why CanonicalKey and ApiKey have the same constructors, (b) why they both accept knowledge of the TypesRegistry.

Ideally _neither_ of them will have the TypesRegistry. We should have an uncanonicalised key, then a component (the schema?) which has the TypesRegistry and produces a canonicalised key.